cmd/ipfs/plugin_linux.go

```diff
@@ -0,0 +1,80 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"plugin"
+
+	node "gx/ipfs/QmUBtPvHKFAX43XMsyxsYpMi3U5VwZ4jYFTo4kFhvAR33G/go-ipld-format"
+)
+
+func init() {
+	LoadPlugins = LoadPluginsLinux
+}
+
+func LoadPluginsLinux(cfgroot string) error {
+	plugdirpath := filepath.Join(cfgroot, "plugins")
+	return filepath.Walk(plugdirpath, func(fi string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			log.Warning("found directory inside plugins directory")
+			return nil
+		}
+
+		if err := loadPlugin(fi); err != nil {
+			return fmt.Errorf("loading plugin %s: %s", fi, err)
+		}
+		return nil
+	})
+}
+
+func loadPlugin(pluginfi string) error {
+	plugin, err := plugin.Open(pluginfi)
+	if err != nil {
+		return err
+	}
+
+	typestri, err := plugin.Lookup("PluginType")
+	if err != nil {
+		return fmt.Errorf("check type: %s", err)
+	}
+
+	typestr, ok := typestri.(*string)
+	if !ok {
+		return fmt.Errorf("'PluginType' var was not a string (got %T)", typestri)
+	}
+
+	switch *typestr {
+	case "ipld":
+		regfunci, err := plugin.Lookup("Register")
+		if err != nil {
+			return fmt.Errorf("ipld plugin had no Register function: %s", err)
+		}
```

Comment:
Regarding API for plugins themselves, I think it might be better to have a plugin expose a structure of callback functions, implemented interface or overridden structure. This way the API is cleanly defined in Go code instead of having to be written somewhere separately.